File: whatsmeow/frontend/src/services/sherlock.ts
import { api } from './api';
import type { Scrape, Lead } from '../types';

export interface ExtractResponse {
  scrape_id: string;
  status: string;
  message: string;
}

export const sherlockService = {
  // Inicia uma campanha de raspagem — retorna scrape_id imediatamente
  extract: (data: { keyword: string; location: string; limit?: number }) =>
    api.post<ExtractResponse>('/admin/sherlock/extract', data),

  // Lista todas as campanhas da empresa
  listScrapes: () =>
    api.get<{ scrapes: Scrape[] }>('/admin/sherlock/scrapes'),

  // Retorna status e total_leads de uma campanha específica
  getScrape: (id: string) =>
    api.get<Scrape>(`/admin/sherlock/scrapes/${id}`),

  // Remove uma campanha e todos seus leads
  deleteScrape: (id: string) =>
    api.delete(`/admin/sherlock/scrapes/${id}`),

  // Leads de uma campanha específica
  getLeadsByScrape: (scrapeId: string) =>
    api.get<{ leads: Lead[]; total: number }>(`/admin/leads/scrape/${scrapeId}`),
};

File: whatsmeow/server/routes/sherlock.go
package routes

import (
	"github.com/labstack/echo/v4"
	"github.com/verbeux-ai/whatsmiau/repositories/leads"
	"github.com/verbeux-ai/whatsmiau/repositories/scrapes"
	"github.com/verbeux-ai/whatsmiau/server/controllers"
	"github.com/verbeux-ai/whatsmiau/server/middleware"
	"github.com/verbeux-ai/whatsmiau/services"
	"go.uber.org/zap"
)

func Sherlock(group *echo.Group) {
	scrapeRepo, err := scrapes.NewSQL()
	if err != nil {
		zap.L().Warn("scrapes repo init failed, sherlock routes not registered", zap.Error(err))
		return
	}

	leadRepo, err := leads.NewSQL()
	if err != nil {
		zap.L().Warn("leads repo init failed, sherlock routes not registered", zap.Error(err))
		return
	}

	sherlockService := services.NewSherlockService()
	ctrl := controllers.NewSherlock(sherlockService, scrapeRepo, leadRepo)

	protected := group.Group("", middleware.JWTAuth, middleware.AdminOnly)

	// POST /v1/admin/sherlock/extract — inicia campanha de raspagem (async)
	protected.POST("/extract", ctrl.Extract)

	// GET  /v1/admin/sherlock/scrapes       — lista campanhas da empresa
	// GET  /v1/admin/sherlock/scrapes/:id   — status de uma campanha
	// DELETE /v1/admin/sherlock/scrapes/:id — remove campanha e seus leads
	protected.GET("/scrapes", ctrl.ListScrapes)
	protected.GET("/scrapes/:id", ctrl.GetScrape)
	protected.DELETE("/scrapes/:id", ctrl.DeleteScrape)
}
